<div style="background-color:#D8232A; color: #FFFFFF; padding: 15px; border-radius: 8px; text-align: center;">
  <img src="./logo.png" alt="Banner Image" style="max-width: 100%; height: auto; max-height: 120px; margin-bottom: 10px; border-radius: 4px;">
  <h1 style="margin: 0;"><i>Data Quality Assessment and Courier Shipment Analytics</i></h1>
</div>


<div style="color: #14368B;padding: 12px;">
<h2><i>Objective</i></h2>
<p><i><strong>This project aims to demonstrate data cleaning, exploratory data analysis, and business insights in the courier or postal-services domain. This will use python and its libraries for this purpose</strong></i></p></div>

<div style="color: #14368B;padding: 12px;">
<h3><i>Dataset Overview</i></h3></div>

<div style="color: #14368B;">
    <p ><i><em></em>A real-world courier shipment dataset extracted from DTDC operations across India. This dataset captures the end-to-end flow of parcel logistics, covering sender/receiver details, weight, delivery mode, charges, and status.</em></i></p>
<p><em><i>Each row represents a courier consignment booked via DTDC, including details like:</em></i></p>
    <ul><li><em><i>Pickup and delivery location (city, pincode, state)</em></i></li>
    <li><em><i>Shipment weight (actual, volumetric, chargeable)</em></i></li>
    <li><em><i>Delivery mode and tariff breakdown</em></i></li>
    <li><em><i>Signatures, timestamps, and GST details (where applicable)</em></i></li>
    <li><em><i>Nature of parcel: documents or non-docs</em></i></li>
    </ul></em></i></p></div>

<div style="color: #14368B;">
    <p><em>The data set contains 42 columns</em></p>
<ul>
  <li><strong>Origin:</strong> <em>The city from which the parcel was originally booked or dispatched.</em></li>
  <li><strong>Destination:</strong> <em>The city where the parcel is intended to be delivered.</em></li>
  <li><strong>Pouch No:</strong> <em>A unique internal identifier (often UUID) used for grouping shipments in logistics.</em></li>
  <li><strong>Date:</strong> <em>The date when the shipment was booked, represented in Excel serial number format.</em></li>
  <li><strong>Sender Name:</strong> <em>Full name of the individual or company sending the parcel.</em></li>
  <li><strong>Sender Phone:</strong> <em>Mobile or landline contact number of the sender.</em></li>
  <li><strong>Sender Address:</strong> <em>Complete address of the sender for pickup and records.</em></li>
  <li><strong>Sender City:</strong> <em>The city of the sender may duplicate the "Origin" field.</em></li>
  <li><strong>Sender State:</strong> <em>The Indian state in which the sender is located.</em></li>
  <li><strong>Sender Pincode:</strong> <em>The 6-digit postal code (ZIP) of the sender's location.</em></li>
  <li><strong>Sender GSTIN:</strong> <em>GST Identification Number of the sender (if applicable for business shipments).</em></li>
  <li><strong>Total Pieces:</strong> <em>The number of individual physical packages included in the consignment.</em></li>
  <li><strong>Actual Wt:</strong> <em>The actual physical weight (in kilograms) of the parcel(s).</em></li>
  <li><strong>Volumetric Wt:</strong> <em>The weight calculated based on volume dimensions, used for billing purposes.</em></li>
  <li><strong>Chargeable Wt:</strong> <em>The weight used for billing , typically the greater of actual or volumetric weight.</em></li>
  <li><strong>Paperwork:</strong> <em>Indicates whether paperwork was submitted along with the parcel (Yes/No).</em></li>
  <li><strong>Sender Signature:</strong> <em>Denotes whether the sender signed the dispatch documentation (Yes/No).</em></li>
  <li><strong>Sender Date:</strong> <em>The date the sender completed handoff or signed documents, in Excel serial format.</em></li>
  <li><strong>Recipient Name:</strong> <em>The full name of the person or organization meant to receive the parcel.</em></li>
  <li><strong>Recipient Phone:</strong> <em>Contact number of the intended recipient.</em></li>
  <li><strong>Recipient Address:</strong> <em>The complete delivery address where the parcel should be sent.</em></li>
  <li><strong>Recipient City:</strong> <em>The city of the recipient may duplicate the "Destination" field.</em></li>
  <li><strong>Receiver State:</strong> <em>The Indian state where the recipient resides.</em></li>
  <li><strong>Receiver Pincode:</strong> <em>The 6-digit postal or ZIP code of the delivery address.</em></li>
  <li><strong>Description:</strong> <em>A short textual description of the parcel contents (e.g., documents, items).</em></li>
  <li><strong>Value Added Services:</strong> <em>Optional services added to the shipment, such as fragile handling, express delivery, insurance, etc.</em></li>
  <li><strong>Consignment No:</strong> <em>The official DTDC tracking number assigned to the shipment.</em></li>
  <li><strong>Expiry Date:</strong> <em>The last valid date for delivery, after which the shipment may expire or be returned. Stored as an Excel serial number.</em></li>
  <li><strong>Booking Code:</strong> <em>The unique code identifying the DTDC branch or booking counter.</em></li>
  <li><strong>Recipient GSTIN:</strong> <em>GST Identification Number of the recipient (if applicable).</em></li>
  <li><strong>Receiver Name:</strong> <em>An alternate or duplicate field for the recipient name.</em></li>
  <li><strong>Relationship:</strong> <em>Describes the relationship of the receiver to the original recipient (e.g., Self, Staff, Family).</em></li>
  <li><strong>Company Stamp:</strong> <em>Indicates whether a company stamp was used during the receipt (Yes/No).</em></li>
  <li><strong>Receiver Signature:</strong> <em>Indicates whether the recipient signed the delivery confirmation (Yes/No).</em></li>
  <li><strong>Receive Date:</strong> <em>The date the parcel was received, in Excel serial format.</em></li>
  <li><strong>Tariff:</strong> <em>The base delivery charge in Indian Rupees.</em></li>
  <li><strong>VAS Charges:</strong> <em>Additional charges for any value-added services applied to the shipment.</em></li>
  <li><strong>Total Amount:</strong> <em>The total amount billed for the consignment, including tariff and any add-ons.</em></li>
  <li><strong>Mode:</strong> <em>The mode of shipment chosen :Surface, Air, or Express.</em></li>
  <li><strong>Risk Surcharge:</strong> <em>Denotes who bears risk liability, either the Owner (sender) or Carrier (DTDC).</em></li>
  <li><strong>Mode of Payment:</strong> <em>The payment method used at booking, such as Cash, Card, UPI, Wallet, etc.</em></li>
  <li><strong>Nature of Consignment:</strong> <em>Indicates the type of parcel "Dox" for documents or "Non-Dox" for goods/merchandise.</em></li>
</ul>
</div>

<div  style="background-color:#D8232A; color: #FFFFFF;padding: 5px; border-radius: 1px;">
<h3><i><em>Import libraries</em></i></h3> 
</div>

In [192]:
import pandas as pd
import numpy as np

<div  style="background-color:#D8232A; color: #FFFFFF;padding: 5px; border-radius: 1px;">
<h3><i><em>Load dataset and basic review</em></i></h3> 
</div>

In [193]:
df = pd.read_csv("Dataset_Generator_for_DTDC.csv")

In [194]:
df.shape # gives the number of rows and columns in the dataset as a tuple

(49639, 42)

<div style="color: #14368B;">
<ul>
    <li><i>.info() gives the column wise Name, non null value count and the Data type.</i></li>
    <p><i>This is the step where we get the highlevel information about:</i></p> 
    <li><i>The column data type</i></li>
    <li><i>Which columns have null values , since from the above .shape we know the row count</i></li>
</ul></div>

In [195]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 49639 entries, 0 to 49638
Data columns (total 42 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Origin                 49639 non-null  str    
 1   Destination            49639 non-null  str    
 2   Pouch No               49639 non-null  str    
 3   Date                   49639 non-null  str    
 4   Sender's Name          49639 non-null  str    
 5   Sender Phone           49639 non-null  int64  
 6   Sender Address         49639 non-null  str    
 7   Sender City            49639 non-null  str    
 8   Sender State           49639 non-null  str    
 9   Sender Pincode         49639 non-null  int64  
 10  Sender GSTIN           26778 non-null  str    
 11  Total Pieces           49639 non-null  int64  
 12  Actual Wt              49639 non-null  float64
 13  Volumetric Wt          49639 non-null  float64
 14  Chargeable Wt          49639 non-null  float64
 15  Paperwork    

In [196]:
df.head() # this gives the first 5 rows of the data frame, which will give an overview of data content in each column

,Origin,Destination,Pouch No,Date,Sender's Name,Sender Phone,Sender Address,Sender City,Sender State,Sender Pincode,...,Company Stamp,Receiver Signature,Receive Date,Tariff,VAS Charges,Total Amount,Mode,Risk Surcharge,Mode of Payment,Nature of Consignment
0,Nagpur,Jamshedpur,55e275ad-8da9-49c9-986c-93134fef2c19,2025-06-25,Megha Majumdar,6137840452,"H.No. 425 Varughese Chowk, Above Shop, Nagpur,...",Nagpur,Maharashtra,264582,...,NaN,NaN,2025-06-30,147.22,70.09,217.31,Express,Carrier,Cash,Dox
1,Jaipur,Vizag,7fab4175-11b2-4090-a0ed-4571ccd68878,2025-06-27,Nanda-Saraf,52907142,"48/880, Tripathi, Above Shop, Jaipur, Rajastha...",Jaipur,Rajasthan,200155,...,NaN,NaN,2025-06-28,256.83,82.20,339.03,Surface,Carrier,Card,Dox
2,Lucknow,Bhubaneswar,5231428d-b524-433e-8043-45c7e39a94aa,2025-06-27,Contractor LLC,5081512738,"257, Kulkarni Path, Above Shop, Lucknow, Uttar...",Lucknow,Uttar Pradesh,353311,...,NaN,NaN,2025-06-28,432.45,107.63,540.08,Surface,Carrier,Cash,Non-Dox
3,Coimbatore,Mumbai,2f556b38-298d-4e8c-9310-4e6a3a5fbe61,2025-06-11,Krishna Hans,1518559754,"83/17 Sridhar Zila, Above Shop, Coimbatore, Ta...",Coimbatore,Tamil Nadu,553663,...,NaN,NaN,2025-06-15,1162.00,25.23,1187.23,Express,Carrier,Card,Non-Dox
4,Meerut,Vijayawada,c3154041-34eb-435e-94d8-0900e6a9ed01,2025-07-01,Kuruvilla-Tiwari,1666398391,"16/64 Jayaraman Circle, Above Shop, Meerut, Ut...",Meerut,Uttar Pradesh,273143,...,NaN,NaN,2025-07-05,285.55,67.70,353.25,Surface,Carrier,Cash,Dox


In [197]:
df.tail() # this gives the last 5 rows of the data frame

,Origin,Destination,Pouch No,Date,Sender's Name,Sender Phone,Sender Address,Sender City,Sender State,Sender Pincode,...,Company Stamp,Receiver Signature,Receive Date,Tariff,VAS Charges,Total Amount,Mode,Risk Surcharge,Mode of Payment,Nature of Consignment
49634,Varanasi,Bhopal,78ad9246-96ca-40e0-a77e-1f5f601b2114,2025-07-04,Chavvi Kashyap,6657295176,"H.No. 66, Chaudry Marg, Behind School, Varanas...",Varanasi,Uttar Pradesh,864652,...,NaN,NaN,2025-07-05,485.50,31.14,516.64,Surface,Carrier,Card,Non-Dox
49635,Patna,Jamshedpur,be889bde-61ea-460c-8dc8-6bdaf7a91579,2025-06-15,"Chakrabarti, Pandit and Agrawal",5273847382,"62 Palla, Opposite Mall, Patna, Bihar - 114677",Patna,Bihar,114677,...,NaN,NaN,2025-06-18,451.46,124.68,576.14,Surface,Carrier,Cash,Dox
49636,Kolkata,Aurangabad,17a77c5d-ccdf-42c1-adc0-ec5b89b84154,2025-06-06,Arya Shere,4416278858,"H.No. 959, Shetty Ganj, Opposite Mall, Kolkata...",Kolkata,West Bengal,19999,...,NaN,NaN,2025-06-08,458.24,63.33,521.57,Surface,Carrier,Cash,Dox
49637,Srinagar,Aurangabad,9eae5dad-b5c0-4047-b1fe-a340dbc4e41e,2025-07-05,Harita Dhaliwal,5568995605,"960, Saini Street, Behind School, Srinagar, Ja...",Srinagar,Jammu & Kashmir,138785,...,Yes,Yes,2025-07-10,515.65,0.00,515.65,Air Cargo,Carrier,Wallet,Non-Dox
49638,Ahmedabad,Amritsar,fe45d57f-d3ab-4c39-b60a-8abebae5c224,2025-06-24,Warhi Nath,2408553775,"13, Manne Ganj, Behind School, Ahmedabad, Guja...",Ahmedabad,Gujarat,587107,...,NaN,NaN,2025-06-29,453.02,106.62,559.64,Surface,Carrier,Card,Dox


<div  style="background-color:#D8232A; color: #FFFFFF;padding: 5px; border-radius: 1px;">
<h3><i><em>Data Cleaning</em></i></h3> 
</div>

<div style="color: #14368B;">
    <p><i>Removing the columns that are not required for analysis.</i></p>
<ul>
    <li><i>"Sender Phone","Sender's Name","Sender Address","Recipient Name","Recipient Phone","Recipient Address","Sender Pincode","Receiver Pincode"</i></li>
    <p><i>The above are removed as these are either personal identifiers and , or repeated data that doesn't help in analysis.</i></p> 
</ul></div>

In [198]:
df = df.drop(columns = ["Sender Phone","Sender's Name","Sender Address","Recipient Name","Recipient Phone","Recipient Address","Sender Pincode","Receiver Pincode"])

In [199]:
df.shape # checking the coulmn count after deletion of columns.

(49639, 34)

In [200]:
df.describe().round(3) # the .descirbe function gives the statistical distribution of numberical coulmns in the df.

,Total Pieces,Actual Wt,Volumetric Wt,Chargeable Wt,Tariff,VAS Charges,Total Amount
count,49639.000,49639.000,49639.000,49639.000,49639.000,49639.000,49639.000
mean,2.994,6.167,7.717,7.717,440.854,67.199,508.053
std,1.415,3.375,3.477,3.477,243.188,47.899,248.097
min,1.000,0.300,0.440,0.440,19.900,0.000,22.630
25%,2.000,3.260,4.800,4.800,255.040,27.305,320.395
50%,3.000,6.190,7.740,7.740,409.980,67.450,477.640
75%,4.000,9.080,10.640,10.640,571.570,108.860,644.805
max,5.000,12.000,14.970,14.970,1372.900,150.000,1481.160


<div style="color: #14368B;">
    <p><i>Converting data types of date columns which are in string to datetime</i></p>
<ul>
    <li><i>"Date","Receive Date","Sender Date","Expiry Date"</i></li>
</ul></div>

In [201]:
df["Date"] = pd.to_datetime(df["Date"])
df["Receive Date"] = pd.to_datetime(df["Receive Date"])
df["Sender Date"] = pd.to_datetime(df["Sender Date"])
df["Expiry Date"] = pd.to_datetime(df["Expiry Date"])

In [202]:
df.dtypes # this will confirm the dtype changes

Origin                              str
Destination                         str
Pouch No                            str
Date                     datetime64[us]
Sender City                         str
Sender State                        str
Sender GSTIN                        str
Total Pieces                      int64
Actual Wt                       float64
Volumetric Wt                   float64
Chargeable Wt                   float64
Paperwork                           str
Sender Signature                    str
Sender Date              datetime64[us]
Recipient City                      str
Receiver State                      str
Description                         str
Value Added Services                str
Consignment No                      str
Expiry Date              datetime64[us]
Booking Code                        str
Recipient GSTIN                     str
Receiver Name                       str
Relationship                        str
Company Stamp                       str


<div style="color: #14368B;">
    <p><i>Filling missing values:</i></p>
    <ul>
        <li><i>Sender GSTIN : Column will have GSTIN number for sender's who are company (not individual).</i></li>
        <li><i>Sender Signature : Indicates if the sender has signed the documents.</i></li>
        <li><i>Value Added Services : Shows the additional services taken by sender.</i></li>
        <li><i>Recipient GSTIN :  Column will have GSTIN number for Recipient's who are company (not individual).</i></li>
        <li><i>Company Stamp : Column will have stamp for Recipient's who are company</i></li>
        <li><i>Receiver Signature :Indicates if the Receiver has signed the delivery confirmation.</i></li>
    </ul>
    <p>Doing cross tab comparission of Sender GSTIN and Sender Signature, this will show relation between the two columns</p>
    <p>This shows that the sender signature is always there where sender GSTIN present,and checking the unique value we get the column has Yes or null, so this is a Yes/no column. We can fill the null with 'No' and since GST is the unque ID , we fill the null with No assuming they are 'individual' senders</p>
     <p>The same is the case for Receiver Signature , Recipient GSTIN</p>
    <p>Company Stamp is present for Receiver Signature but not mandatory. So we can fill null with 'No'</p>
</div>

In [204]:
df["Sender Signature"].isna().sum()

np.int64(19828)

In [205]:
print(df['Sender Signature'].unique())
print(df['Receiver Signature'].unique())
print(df['Company Stamp'].unique())

<ArrowStringArray>
[nan, 'Yes']
Length: 2, dtype: str
<ArrowStringArray>
[nan, 'Yes']
Length: 2, dtype: str
<ArrowStringArray>
[nan, 'Yes']
Length: 2, dtype: str


In [206]:
pd.crosstab(df["Sender GSTIN"].isna(),df["Sender Signature"].isna())

Sender Signature,False,True
Sender GSTIN,,
False,26778,0
True,3033,19828


In [207]:
pd.crosstab(df['Company Stamp'].isna(), df['Receiver Signature'].isna())

Receiver Signature,False,True
Company Stamp,,
False,19865,0
True,0,29774


In [208]:
df.loc[df["Value Added Services"].isna(),"VAS Charges"].sum()

np.float64(0.0)

<div style="color: #14368B;">
<p>The above checks if Value Added Services has null where the VAS charges are zero.</p>
<p>Since the sum is zero, we can conclude that the sender didn't choose any additional services for parcels, hence we can fill with 'None'. </p>
</div>

In [ ]:
<div style="color: #14368B;">
<p>Filling the missing values using .fillna</p>
</div>

In [209]:
df = df.fillna({"Sender Signature":"NO","Receiver Signature":"No","Company Stamp":"No","Sender GSTIN":"Not applied","Recipient GSTIN":"Not applied","Value Added Services":"None"})

In [211]:
df[['Date', 'Sender Date', 'Receive Date', 'Expiry Date']].info()

<class 'pandas.DataFrame'>
RangeIndex: 49639 entries, 0 to 49638
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          49639 non-null  datetime64[us]
 1   Sender Date   49639 non-null  datetime64[us]
 2   Receive Date  49639 non-null  datetime64[us]
 3   Expiry Date   49639 non-null  datetime64[us]
dtypes: datetime64[us](4)
memory usage: 1.5 MB


In [213]:
(df['Date'] == df['Sender Date']).all() # checking if the Date column and Sender date is the same

np.True_

<div style="color: #14368B;">
<p>Since both columns are same we drop the date column</p>
</div>

In [216]:
df = df.drop(columns = 'Date')

<div style="color: #14368B;">
<p>Error checking if any date have unrealistic values</p>
</div>

In [214]:
(df['Sender Date'] > df['Receive Date']).sum() # All sender date are less than reciever date

np.int64(0)

In [218]:
(df['Sender Date'] > df['Expiry Date']).sum()  # All sender date are less than expiry date

np.int64(0)

<div style="color: #14368B;">
<p><i>Duplicate Handling</i></p>
</div>

In [222]:
df.duplicated().sum()

np.int64(0)

In [223]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 49639 entries, 0 to 49638
Data columns (total 33 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Origin                 49639 non-null  str           
 1   Destination            49639 non-null  str           
 2   Pouch No               49639 non-null  str           
 3   Sender City            49639 non-null  str           
 4   Sender State           49639 non-null  str           
 5   Sender GSTIN           49639 non-null  str           
 6   Total Pieces           49639 non-null  int64         
 7   Actual Wt              49639 non-null  float64       
 8   Volumetric Wt          49639 non-null  float64       
 9   Chargeable Wt          49639 non-null  float64       
 10  Paperwork              49639 non-null  str           
 11  Sender Signature       49639 non-null  str           
 12  Sender Date            49639 non-null  datetime64[us]
 13  Recipient Ci

<div style="color: #14368B;">
<p><i>Outlier checking</i></p>
</div>

In [224]:
df[['Actual Wt', 'Volumetric Wt', 'Chargeable Wt', 'Tariff', 'VAS Charges', 'Total Amount']].describe()

,Actual Wt,Volumetric Wt,Chargeable Wt,Tariff,VAS Charges,Total Amount
count,49639.000000,49639.000000,49639.000000,49639.000000,49639.000000,49639.000000
mean,6.167263,7.716568,7.716568,440.854079,67.198937,508.053016
std,3.375089,3.476748,3.476748,243.188225,47.899430,248.097236
min,0.300000,0.440000,0.440000,19.900000,0.000000,22.630000
25%,3.260000,4.800000,4.800000,255.040000,27.305000,320.395000
50%,6.190000,7.740000,7.740000,409.980000,67.450000,477.640000
75%,9.080000,10.640000,10.640000,571.570000,108.860000,644.805000
max,12.000000,14.970000,14.970000,1372.900000,150.000000,1481.160000


<div style="color: #14368B;">
    <p><i>The mean and median values of all columns are almost same and comparing all values it seems to have an evenly distributed data.</i></p>
    <p><i>Only the tariff data seems to have a high max value almost double but that could be geniune large parcel.</i></p>
    <p><i>We can check that with IQR method </i></p>
</div>

In [237]:
Q1 = df["Tariff"].quantile(0.25)
Q3 = df["Tariff"].quantile(0.75)
IQR = Q3-Q1
lowerBand = Q1 - 1.5 * IQR
upperBand = Q3 + 1.5 * IQR
print(f"Upper limit is {upperBand}")
print(f"Lower limit is {lowerBand}")
outliers = df[(df['Tariff'] < lowerBand) | (df['Tariff'] > upperBand)]
print(outliers.shape)

Upper limit is 1046.3649999999998
Lower limit is -219.75499999999997
(1112, 33)


<div style="color: #14368B;">
    <p><i>Note : since the lower band is negative, no rows are under the lower limit</i></p>
    <p><i>The upper limit of 1112 rows could be actual parcels with high volumetric weight.</i></p>
    <p><i>To confirm that check the .describe of all coulmns for tariff greater than the upper limit alone</i></p>
</div>

In [247]:
df[df["Tariff"]>1046.36].sort_values(by ="Tariff",ascending  = False)[['Chargeable Wt', 'Mode', 'Origin', 'Destination', 'Total Amount']]

,Chargeable Wt,Mode,Origin,Destination,Total Amount
9020,14.84,Express,Mumbai,Delhi,1372.90
18098,14.73,Express,Coimbatore,Amritsar,1383.90
18391,14.76,Express,Patna,Nagpur,1370.19
3136,14.62,Express,Raipur,Bangalore,1407.41
7223,14.42,Express,Mumbai,Srinagar,1481.16
...,...,...,...,...,...
47654,12.98,Express,Jaipur,Shimla,1144.21
24086,13.08,Express,Dehradun,Mumbai,1165.29
30196,11.85,Express,Raipur,Varanasi,1046.47
35405,13.31,Express,Jamshedpur,Shimla,1046.44


<div style="color: #14368B;">
    <p><i>Standarizing column content</i></p>
</div>

In [248]:
df.columns

Index(['Origin', 'Destination', 'Pouch No', 'Sender City', 'Sender State',
       'Sender GSTIN', 'Total Pieces', 'Actual Wt', 'Volumetric Wt',
       'Chargeable Wt', 'Paperwork', 'Sender Signature', 'Sender Date',
       'Recipient City', 'Receiver State', 'Description',
       'Value Added Services', 'Consignment No', 'Expiry Date', 'Booking Code',
       'Recipient GSTIN', 'Receiver Name', 'Relationship', 'Company Stamp',
       'Receiver Signature', 'Receive Date', 'Tariff', 'VAS Charges',
       'Total Amount', 'Mode', 'Risk Surcharge', 'Mode of Payment',
       'Nature of Consignment'],
      dtype='str')

In [252]:
columns = ['Origin', 'Destination', 'Pouch No', 'Sender City', 'Sender State',
        'Paperwork', 'Sender Signature','Recipient City', 'Receiver State', 'Description',
       'Value Added Services','Receiver Name', 'Relationship', 'Company Stamp',
       'Receiver Signature',  'Mode', 'Risk Surcharge', 'Mode of Payment','Nature of Consignment']
for col in columns:
    df[col] = df[col].str.strip().str.title()


In [258]:
allcol = ['Origin', 'Destination', 'Pouch No', 'Sender City', 'Sender State',
       'Sender GSTIN', 'Total Pieces', 'Actual Wt', 'Volumetric Wt',
       'Chargeable Wt', 'Paperwork', 'Sender Signature', 'Sender Date',
       'Recipient City', 'Receiver State', 'Description',
       'Value Added Services', 'Consignment No', 'Expiry Date', 'Booking Code',
       'Recipient GSTIN', 'Receiver Name', 'Relationship', 'Company Stamp',
       'Receiver Signature', 'Receive Date', 'Tariff', 'VAS Charges',
       'Total Amount', 'Mode', 'Risk Surcharge', 'Mode of Payment',
       'Nature of Consignment']
for col in allcol:
    print(df[col].unique())

<ArrowStringArray>
[            'Nagpur',             'Jaipur',            'Lucknow',
         'Coimbatore',             'Meerut',              'Delhi',
          'Ahmedabad',            'Chennai',           'Dehradun',
        'Bhubaneswar',         'Chandigarh',              'Vizag',
         'Vijayawada',             'Nashik',           'Srinagar',
          'Hyderabad',            'Kolkata',           'Amritsar',
              'Surat',             'Ranchi',           'Ludhiana',
           'Guwahati',              'Patna',             'Mumbai',
             'Bhopal',              'Kochi',             'Raipur',
             'Indore',             'Shimla',         'Jamshedpur',
         'Aurangabad',          'Bangalore', 'Thiruvananthapuram',
           'Varanasi',               'Agra',               'Pune']
Length: 36, dtype: str
<ArrowStringArray>
[        'Jamshedpur',              'Vizag',        'Bhubaneswar',
             'Mumbai',         'Vijayawada',             'Nagpur',
 

In [ ]:
df.to_excel("Cleaned_DTDC_Data.xlsx", index=False)